runners/direct-java/src/main/java/org/apache/beam/runners/direct/UnboundedReadEvaluatorFactory.java

```diff
@@ -146,35 +170,87 @@ public void processElement(WindowedValue<Object> element) {}
     @Override
     public InProcessTransformResult finishBundle() throws IOException {
       UncommittedBundle<OutputT> output = evaluationContext.createRootBundle(transform.getOutput());
-      try (UnboundedReader<OutputT> reader =
-              createReader(source, evaluationContext.getPipelineOptions());) {
-        int numElements = 0;
-        if (reader.start()) {
+      try {
+        if (checkpointMark != null) {
+          checkpointMark.finalizeCheckpoint();
+        }
+
+        boolean elementAvailable = startReader();
+
+        Instant watermark = currentReader.getWatermark();
+        if (elementAvailable) {
+          int numElements = 0;
           do {
-            output.add(
-                WindowedValue.timestampedValueInGlobalWindow(
-                    reader.getCurrent(), reader.getCurrentTimestamp()));
+            output.add(WindowedValue.timestampedValueInGlobalWindow(currentReader.getCurrent(),
+                currentReader.getCurrentTimestamp()));
             numElements++;
-          } while (numElements < ARBITRARY_MAX_ELEMENTS && reader.advance());
+          } while (numElements < ARBITRARY_MAX_ELEMENTS && currentReader.advance());
+          watermark = currentReader.getWatermark();
+          // Only take a checkpoint if we did any work
+          finishRead();
         }
```

Comment:
move finalize down here? We don't want users somehow relying on finalize happening immediately.